demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -43,88 +44,503 @@
+{
+    CK_RV xResult;
+    CK_BBOOL xTrue = CK_TRUE;
+    CK_FUNCTION_LIST_PTR pxFunctionList;
+    CK_OBJECT_CLASS xClass = CKO_PUBLIC_KEY;
+
+    xResult = C_GetFunctionList( &pxFunctionList );
+
+    int lMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext;
+
+    mbedtls_pk_init( &xMbedPkContext );
+    lMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, pucKey, xKeyLength, NULL, 0 );
+
+    if( lMbedResult != 0 )
+    {
+        lMbedResult = mbedtls_pk_parse_public_key( &xMbedPkContext, pucKey, xKeyLength );
+    }
+
+    if( xPublicKeyType == CKK_RSA )
+    {
+        CK_BYTE xPublicExponent[] = { 0x01, 0x00, 0x01 };
+        CK_BYTE xModulus[ MODULUS_LENGTH + 1 ] = { 0 };
+
+        lMbedResult = mbedtls_rsa_export_raw( ( mbedtls_rsa_context * ) xMbedPkContext.pk_ctx,
+                                              ( unsigned char * ) &xModulus, MODULUS_LENGTH + 1,
+                                              NULL, 0,
+                                              NULL, 0,
+                                              NULL, 0,
+                                              NULL, 0 );
+        CK_ATTRIBUTE xPublicKeyTemplate[] =
+        {
+            { CKA_CLASS,           &xClass,           sizeof( CK_OBJECT_CLASS )                    },
+            { CKA_KEY_TYPE,        &xPublicKeyType,   sizeof( CK_KEY_TYPE )                        },
+            { CKA_TOKEN,           &xTrue,            sizeof( xTrue )                              },
+            { CKA_MODULUS,         &xModulus + 1,     MODULUS_LENGTH                               }, /* Extra byte allocated at beginning for 0 padding. */
+            { CKA_VERIFY,          &xTrue,            sizeof( xTrue )                              },
+            { CKA_PUBLIC_EXPONENT, xPublicExponent,   sizeof( xPublicExponent )                    },
+            { CKA_LABEL,           pucPublicKeyLabel, strlen( ( const char * ) pucPublicKeyLabel ) }
+        };
+
+        xResult = pxFunctionList->C_CreateObject( xSession,
+                                                  ( CK_ATTRIBUTE_PTR ) xPublicKeyTemplate,
+                                                  sizeof( xPublicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                  pxPublicKeyHandle );
+    }
+    else if( xPublicKeyType == CKK_EC )
+    {
+        CK_BYTE xEcParams[] = pkcs11DER_ENCODED_OID_P256;
+        size_t xLength;
+        CK_BYTE xEcPoint[ 256 ] = { 0 };
+
+        mbedtls_ecdsa_context * pxEcdsaContext;
+        pxEcdsaContext = ( mbedtls_ecdsa_context * ) xMbedPkContext.pk_ctx;
+
+        /* DER encoded EC point. Leave 2 bytes for the tag and length. */
+        lMbedResult = mbedtls_ecp_point_write_binary( &pxEcdsaContext->grp, &pxEcdsaContext->Q,
+                                                      MBEDTLS_ECP_PF_UNCOMPRESSED, &xLength,
+                                                      xEcPoint + 2, sizeof( xEcPoint ) - 2 );
```

Comment:
Does this agree with the comment?  Seems like a gap of 4 based on the last two arguments.

Reply:
This comment is accurate.

https://tls.mbed.org/api/ecp_8h.html#af7d35a6631f662cc279eed8c4f0f7ba6

The second to last argument xEcPoint+2 is the pointer to the location where the data is being copied (so leaving 2 bytes at the beginning for tag and length).   The last argument sizeof(xEcPoint)-2 is the length of the data being copied, which is 2 less than the full buffer length to account for the 2 bytes for length and tag which are written on the subsequent 2 lines.